consensus/src/processes/parents_builder.rs

```diff
@@ -119,7 +115,7 @@ impl<T: HeaderStoreReader, U: ReachabilityStoreReader, V: RelationsStoreReader>
                         reference_blocks
                     };
 
-                    if is_empty_level {
+                    if candidates_by_level_to_reference_blocks_map[block_level].is_empty() {
```

Comment:
This doesn't look equivalent to me, since the current loop mutates `candidates_by_level_to_reference_blocks_map`

Reply:
Seems like all we can optimize here is the case of no direct parents at level + first parent has more than one parent at level -- a case which should be very rare (since low levels are expected to have direct parents, and high levels are a chain)